internal/hcs/system.go

```diff
@@ -326,13 +335,108 @@ func (computeSystem *System) Properties(ctx context.Context, types ...schema1.Pr
+				MemoryUsagePrivateWorkingSetBytes: privateWorkingSet,
+			},
+			Processor: &hcsschema.ProcessorStats{
+				RuntimeKernel100ns: uint64(processorInfo.TotalKernelTime),
+				RuntimeUser100ns:   uint64(processorInfo.TotalUserTime),
+				TotalRuntime100ns:  uint64(processorInfo.TotalKernelTime + processorInfo.TotalUserTime),
+			},
+			Storage: &hcsschema.StorageStats{
+				ReadCountNormalized:  uint64(storageInfo.ReadStats.IoCount),
+				ReadSizeBytes:        storageInfo.ReadStats.TotalSize,
+				WriteCountNormalized: uint64(storageInfo.WriteStats.IoCount),
+				WriteSizeBytes:       storageInfo.WriteStats.TotalSize,
+			},
+		},
+	}, nil
+}
+
 // PropertiesV2 returns the requested container properties targeting a V2 schema container.
 func (computeSystem *System) PropertiesV2(ctx context.Context, types ...hcsschema.PropertyType) (*hcsschema.Properties, error) {
 	computeSystem.handleLock.RLock()
 	defer computeSystem.handleLock.RUnlock()
 
 	operation := "hcs::System::PropertiesV2"
 
+	// There's an optimization we can make here if the user is asking for Statistics only, and this is due to
+	// an inefficient way that HCS calculates the private working set total for a given container. HCS
+	// calls NtQuerySystemInformation with the class SystemProcessInformation which returns an array containing
+	// system information for *every* process running on the machine. They then grab the pids that are running in
+	// the container and filter down the entries in the array to only what's running in that silo and start tallying
+	// up the total. This doesn't work well as performance should get worse if more processess are running on the
+	// machine in general and not just in the container. All of the additional information besides the
+	// WorkingSetPrivateSize field is ignored as well which isn't great and is wasted work to fetch.
+	//
+	// HCS only let's you grab statistics in an all or nothing fashion, so we can't just grab the private
+	// working set ourselves and ask for everything else seperately. The optimization we can make here is
+	// to open the silo ourselves and do the same queries for the rest of the info, as well as calculating
+	// the private working set in a more efficient manner by:
+	//
+	// 1. Find the pids running in the silo
+	// 2. Get a process handle for every process (only need PROCESS_QUERY_LIMITED_INFORMATION access)
+	// 3. Call NtQueryInformationProcess on each process with the class ProcessVmCounters
+	// 4. Tally up the total using the field PrivateWorkingSetSize in VM_COUNTERS_EX2.
+	//
+	// The only caveat is the process invoking this code must be running as SYSTEM to open the silo,
+	// otherwise we'll fallback to asking HCS for the information.
+
+	// First check if we're only asking for Stats to see if we can apply the above optimization. We should
+	// only do this for containers and not UVMs.
+	if (len(types) == 1 && types[0] == hcsschema.PTStatistics) && computeSystem.typ == "container" {
```

Comment:
Even if they ask for e.g. `PTStatistics` and `PTMemory`, can't we still use the fast path for the `PTStatistics` part and only call HCS for `PTMemory`?

Reply:
@kevpar Thinking about this more, would you be fine with a follow up that adds this? We can actually do the ProcessList query (and memory iirc) entirely in the shim as well which would shake up how I have this structured at the moment. Right now it's simple to reason with but having > 1 query being able to be serviced by the shim and handling the fallback switches it up a bit.

Reply:
Can we just avoid calling HCS for `PTStatistics` for now, and reimplement the other types later?

Reply:
Avoid calling HCS if statistics isn't the only query in the set you mean?

Reply:
I mean that we should never issue a HCS query for `PTStatistics` anymore, since we can entirely fetch this data from code in the shim instead. We should still issue queries to HCS for other property types.

e.g.:

If request is for `[ PTStatistics ]`, we should:
- Create a `Properties` struct (`p`)
- Set the general top-level members (`p.Id`, `p.Owner`, etc.)
- Call `statisticsInProc` and assign its result to `p.Statistics`
- Return `p`

If request is for `[ PTMemory ]`, we should:
- Create a `Properties` struct (`p`)
- Set the general top-level members (`p.Id`, `p.Owner`, etc.)
- Call `PropertiesV2` with `[ PTMemory ]` and assign its `Memory` to `p.Memory`
- Return `p`

If request is for `[ PTStatistics, PTMemory ]`, we should:
- Create a `Properties` struct (`p`)
- Set the general top-level members (`p.Id`, `p.Owner`, etc.)
- Call `statisticsInProc` and assign its result to `p.Statistics`
- Call `PropertiesV2` with `[ PTMemory ]` and assign its `Memory` to `p.Memory`
- Return `p`

Reply:
Wanted to make sure your follow-up didn't sway from the original ask, right I understand what you were getting at originally. My hold off (at least for this PR) is we don't ever query for more than one property in our code for containers. We query once for both memory and stats for UVMs but that's not relevant here as we don't have logic to open the vmwp's job, although from looking the name is just the vms ID haha. Guess its worth future proofing though. I'll add

Reply:
Added. I'd like to push a follow up that makes this nicer when we support more than one query from the shim. Something like:

```go
// queryInProc handles querying for container properties without reaching out to HCS.
func (computeSystem *System) queryInProc(ctx context.Context, props *hcsschema.Properties, types []hcsschema.PropertyType) error {
	// In the future we can make use of some new functionality in the HCS that allows you
	// to pass a job object for HCS to use for the container. Currently, the only way we'll
	// be able to open the job/silo is if we're running as SYSTEM.
	jobOptions := &jobobject.Options{
		UseNTVariant: true,
		Name:         siloNameFmt(computeSystem.id),
	}
	job, err := jobobject.Open(ctx, jobOptions)
	if err != nil {
		return err
	}
	defer job.Close()

	// Grab the stats one by one. If a query was successful remove it from the list so we don't requery
	// the same info from HCS (if we even need to).
	for _, prop := range types {
		switch prop {
		case hcsschema.PTStatistics:
			props.Statistics, err = computeSystem.statisticsInProc(ctx, job)
			if err != nil {
				return err
			}
			// Statistics fills in owner as well.
			props.Owner = computeSystem.owner
                 case hcsschema.PTProcessList:
			props.Statistics, err = computeSystem.processListInProc(ctx, job)
			if err != nil {
				return err
			}
		default:
			log.G(ctx).WithField("propertyType", prop).Warn("Unknown property query")
		}
	}
	return nil
}
```


I don't think that method should return on the first error either, but just showing as a concept